drivers/usb/udc/udc_dwc2.c

```diff
@@ -674,22 +674,35 @@ static void dwc2_handle_enumdone(const struct device *dev)
 	priv->enumspd = usb_dwc2_get_dsts_enumspd(dsts);
 }
 
-static inline int dwc2_read_fifo_setup(const struct device *dev)
+static inline int dwc2_read_fifo_setup(const struct device *dev, uint8_t ep,
+				       const size_t bytes)
 {
 	struct usb_dwc2_reg *const base = dwc2_get_base(dev);
 	struct udc_dwc2_data *const priv = udc_get_private(dev);
+	size_t offset;
 
 	/* FIFO access is always in 32-bit words */
 
+	if (bytes != 8) {
+		LOG_ERR("%d bytes SETUP", bytes);
+	}
+
 	/*
 	 * We store the setup packet temporarily in the driver's private data
 	 * because there is always a race risk after the status stage OUT
 	 * packet from the host and the new setup packet. This is fine in
 	 * bottom-half processing because the events arrive in a queue and
 	 * there will be a next net_buf for the setup packet.
 	 */
-	sys_put_le32(sys_read32(UDC_DWC2_EP_FIFO(base, 0)), priv->setup);
-	sys_put_le32(sys_read32(UDC_DWC2_EP_FIFO(base, 0)), &priv->setup[4]);
+	for (offset = 0; offset < MIN(bytes, 8); offset += 4) {
+		sys_put_le32(sys_read32(UDC_DWC2_EP_FIFO(base, ep)),
+			     &priv->setup[offset]);
+	}
+
+	/* On protocol error simply discard extra data */
+	while (offset < bytes) {
+		sys_read32(UDC_DWC2_EP_FIFO(base, ep));
+	}
```

Comment:
Is not something missed here? Maybe do the same as in dwc2_read_fifo()
```
	if (unlikely(size > len)) {
		for (uint32_t n = 0; n < DIV_ROUND_UP(size - len, 4U); n++) {
			(void)sys_read32(UDC_DWC2_EP_FIFO(base, ep));
		}
	}
```
or should dwc2_read_fifo() called to discard extra data?

Reply:
`offset += 4;` is missing inside the loop, will add it (without it it is just an endless loop). I don't think going with `unlikely` inside if is going to bring any real benefits. This is not performance critical code path.

Reply:
"ulikely" is more for readability